Attribute Errors raised in Parser Silently Ignored

### Discussed in https://github.com/encode/django-rest-framework/discussions/9426

<div type='discussions-op-text'>

<sup>Originally posted by **james-mchugh** June  3, 2024</sup>
Hello.

First, thank you for the hard work on DRF. It has been great workin with it.

There seems to be an issue where an `AttributeError` raised in a parser is being swallowed by the `Request` class's `__getattr__` method. I've noticed a similar issue reported previously (#4896), but it was closed as the reporter associated with an error from a separate package.

I encountered this because I wrote the server and tested using Python 3.11, and another developer attempted to test using 3.10. My custom parser used a function that was not available in Python 3.10 (hashlib.file_digest), which caused an `AttributeError` when hitting certain endpoints. Instead of seeing a 500 response and a traceback in the server logs, we were getting a 200 response with an empty dictionary in the response body. We were scratching our heads for a while, as there was no indication anything was going wrong.

### Environment

OS: Darwin 23.4.0 Darwin Kernel Version 23.4.0 x86_64
Python: 3.10 + 3.11
DRF: 3.15.1
Django: 5.0.6

### Expected Behavior

500 Internal server error response and traceback in server logs.

### Actual Behavior

200 response and no errors in server logs.

```shell
$ curl --json '{"foo": "bar"}' http://localhost:8000/foos/
{}
```

```text
[04/Jun/2024 02:59:15] "POST /foos/ HTTP/1.1" 200 2
```



### Code to Reproduce

```python
from rest_framework import viewsets, response, parsers, routers


class BrokenParser(parsers.JSONParser):

    def parse(self, stream, media_type=None, parser_context=None):
        raise AttributeError


class TestViewSet(viewsets.ViewSet):

    parser_classes = (BrokenParser,)

    def create(self, request, **kwargs):
        return response.Response(request.data)

router = routers.DefaultRouter()
router.register("foos", TestViewSet, "foo")

urlpatterns = router.urls
```

### Investigation

This appears to be happening because accessing the `data` property lazily parses data. If the parsing raises an `AttributeError`, this is raised up to https://github.com/encode/django-rest-framework/blob/master/rest_framework/request.py#L359 where `self._full_data` is set to an empty dictionary before the error is re-raised.

This error then raises up and causes the attribute access to fallback to the `Request.__getattr__` method via Python's [data model](https://docs.python.org/3/reference/datamodel.html#object.__getattribute__). Here, the `data` attribute is attempted to be retrieved from the `WSGIRequest` stored in `Request._request`, and once again, this raises an `AttributeError` which is caught by https://github.com/encode/django-rest-framework/blob/master/rest_framework/request.py#L423. From here, the original `Request.__getattribute__` handling runs again. Now when the `data` getter runs, `self._full_data` is already set to an empty dictionary which is then returned.

In the end, an empty response is returned and the error is silently ignored.

Noob question, but why does `Request.__getattr__` call `self.__getattribute__` when an `AttributeError` occurs? Based on the Python Data model linked above, `__getattr__` should only run if `__getattribute__` fails in the first place, so by the time `__getattr__` runs, `__getattribute__` already tried and failed.</div>

Just try it!

```python
class TestBrokenPerser(SimpleTestCase):
    def setUp(self):
        self.factory = APIRequestFactory()

    def test_post_accessed_in_post_method(self):
        django_request = self.factory.post('/', {'foo': 'bar'}, content_type='application/json')
        request = Request(django_request, parsers=[BrokenParser()])
        with self.assertRaises(AttributeError, msg='no in parse'):
            request._parse()

        with self.assertRaises(AttributeError, msg='no in data'):
            request.data
```

> FAILED tests/test_parsers.py::TestBrokenPerser::test_post_accessed_in_post_method - AssertionError: AttributeError not raised : no in data

If you follow the call-stack:

- `Request.data` (property)
- `Request._load_data_and_files` (does not catch)
- `Request._parse` (does not catch)
- `BrokenParser.parse` (raises `AttributeError`)

The `AttributeError` is raised and reaches a property and then `__getattr__` is called, just as explained in that bug.

--- a/rest_framework/request.py
+++ b/rest_framework/request.py
@@ -217,7 +217,8 @@ class Request:
     @property
     def data(self):
         if not _hasattr(self, '_full_data'):
-            self._load_data_and_files()
+            with wrap_attributeerrors():
+                self._load_data_and_files()
         return self._full_data
 
     @property
@@ -420,13 +421,14 @@ class Request:
             _request = self.__getattribute__("_request")
             return getattr(_request, attr)
         except AttributeError:
-            return self.__getattribute__(attr)
+            raise AttributeError(f"'{self.__class__.__name__}' object has no attribute '{attr}'")
 
     @property
     def POST(self):
         # Ensure that request.POST uses our request parsing.
         if not _hasattr(self, '_data'):
-            self._load_data_and_files()
+            with wrap_attributeerrors():
+                self._load_data_and_files()
         if is_form_media_type(self.content_type):
             return self._data
         return QueryDict('', encoding=self._request._encoding)
@@ -437,7 +439,8 @@ class Request:
         # Different from the other two cases, which are not valid property
         # names on the WSGIRequest class.
         if not _hasattr(self, '_files'):
-            self._load_data_and_files()
+            with wrap_attributeerrors():
+                self._load_data_and_files()
         return self._files
 
     def force_plaintext_errors(self, value):

--- a/tests/test_request.py
+++ b/tests/test_request.py
@@ -126,6 +126,25 @@ class TestContentParsing(TestCase):
         request.parsers = (PlainTextParser(), )
         assert request.data == content
 
+    def test_calling_data_fails_when_attribute_error_is_raised(self):
+        """
+        Ensure attribute errors raised when parsing are properly re-raised.
+        """
+        expected_message = "Internal error"
+
+        class BrokenParser:
+            media_type = "application/json"
+
+            def parse(self, *args, **kwargs):
+                raise AttributeError(expected_message)
+
+        http_request = factory.post('/', data={}, format="json")
+        request = Request(http_request)
+        request.parsers = (BrokenParser,)
+
+        with self.assertRaisesMessage(WrappedAttributeError, expected_message):
+            request.data
+
 
 class MockView(APIView):
     authentication_classes = (SessionAuthentication,)

